library/core/src/intrinsics/mod.rs

```diff
@@ -4537,6 +4537,32 @@ pub(crate) const fn miri_promise_symbolic_alignment(ptr: *const (), align: usize
     )
 }
 
+#[requires(crate::mem::size_of::<T>() >= crate::mem::size_of::<U>())] //U cannot be larger than T
+#[ensures(|ret: &U| (ret as *const U as usize) % crate::mem::align_of::<U>() == 0)] //check that the output has expected alignment
+pub unsafe fn transmute_unchecked_wrapper<T,U>(input: T) -> U {    
+    unsafe { transmute_unchecked(input) }
+}
+
+//This requires means [output is char implies input is valid unicode value]
+#[requires(type_name::<U>() != type_name::<char>() || (input <= T::from(0xD7FF) || (input >= T::from(0xE000) && input <= T::from(0x10FFFF)) ))]
+#[ensures(|ret: &U| (ret as *const U as usize) % crate::mem::align_of::<U>() == 0)] 
+pub unsafe fn transmute_unchecked_from_u32<T,U>(input: T) -> U
```

Comment:
I don't think we want to generate type specific wrappers.

Reply:
Thank you for your comments! 

One thing I was wondering, we couldn't find any ways to verify type-specific properties without putting constraints on the input/output types (hence the type-specific wrappers). In that case, would it be better to leave all type-specific validity checks to the "`-Z valid-value-checks`" logic, and to only verify generic properties in the generic wrapper (e.g., `#[ensures(|ret: &U| (ub_checks::can_dereference(ret as *const U)))]`)?